4.14 branch fails to build with GCC 12.1

It looks like the latest release of GCC 12.1 triggers an additional warning that makes the build fail on the 4.14 branch:

```
gcc -c -O2 -fno-strict-aliasing -fwrapv -pthread -Wall -Wdeclaration-after-statement -Werror -fno-common -fexcess-precision=standard -fno-tree-vrp -ffunction-sections -g  -D_FILE_OFFSET_BITS=64 -DCAML_NAME_SPACE  -DCAMLDLLIMPORT=  -o interp.b.o interp.c
interp.c: In function ‘caml_interprete’:
interp.c:253:21: error: storing the address of local variable ‘lbl_ACC0’ in ‘caml_instr_base’ [-Werror=dangling-pointer=]
  253 |     caml_instr_base = Jumptbl_base;
      |                     ^
interp.c:51:26: note: ‘lbl_ACC0’ declared here
   51 | #  define Instruct(name) lbl_##name
      |                          ^~~~
interp.c:323:5: note: in expansion of macro ‘Instruct’
  323 |     Instruct(ACC0):
      |     ^~~~~~~~
In file included from interp.c:25:
caml/fix_code.h:38:15: note: ‘caml_instr_base’ declared here
   38 | extern char * caml_instr_base;
      |               ^~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

--- a/Changes
+++ b/Changes
@@ -415,6 +415,9 @@ OCaml 5.0
   (David Allsopp and Nicolás Ojeda Bär, review by Nicolás Ojeda Bär and
    Sebastien Hinderer)
 
+- #11358, #11379: Refactor the initialization of bytecode threading,
+  This avoids a "dangling pointer" warning of GCC 12.1.
+  (Xavier Leroy, report by Armaël Guéneau, review by Gabriel Scherer)
 
 OCaml 4.14.0
 ----------------

--- a/runtime/caml/fix_code.h
+++ b/runtime/caml/fix_code.h
@@ -34,8 +34,7 @@ void caml_set_instruction (code_t pos, opcode_t instr);
 int caml_is_instruction (opcode_t instr1, opcode_t instr2);
 
 #ifdef THREADED_CODE
-extern char ** caml_instr_table;
-extern char * caml_instr_base;
+void caml_init_thread_code(void ** instr_table, void * instr_base);
 void caml_thread_code (code_t code, asize_t len);
 #endif
 

--- a/runtime/fix_code.c
+++ b/runtime/fix_code.c
@@ -82,8 +82,14 @@ void caml_fixup_endianness(code_t code, asize_t len)
 
 #ifdef THREADED_CODE
 
-char ** caml_instr_table;
-char * caml_instr_base;
+static char ** caml_instr_table;
+static char * caml_instr_base;
+
+void caml_init_thread_code(void ** instr_table, void * instr_base)
+{
+  caml_instr_table = (char **) instr_table;
+  caml_instr_base = (char *) instr_base;
+}
 
 static int* opcode_nargs = NULL;
 int* caml_init_opcode_nargs(void)

--- a/runtime/interp.c
+++ b/runtime/interp.c
@@ -54,9 +54,9 @@ sp is a local copy of the global variable Caml_state->extern_sp. */
 #ifdef THREADED_CODE
 #  define Instruct(name) lbl_##name
 #  if defined(ARCH_SIXTYFOUR) && !defined(ARCH_CODE32)
-#    define Jumptbl_base ((char *) &&lbl_ACC0)
+#    define Jumptbl_base &&lbl_ACC0
 #  else
-#    define Jumptbl_base ((char *) 0)
+#    define Jumptbl_base 0
 #    define jumptbl_base ((char *) 0)
 #  endif
 #  ifdef DEBUG
@@ -270,8 +270,7 @@ value caml_interprete(code_t prog, asize_t prog_size)
       (char *) raise_unhandled_code + sizeof(raise_unhandled_code),
       DIGEST_IGNORE, NULL);
 #ifdef THREADED_CODE
-    caml_instr_table = (char **) jumptable;
-    caml_instr_base = Jumptbl_base;
+    caml_init_thread_code(jumptable, Jumptbl_base);
     caml_thread_code(raise_unhandled_code,
                      sizeof(raise_unhandled_code));
 #endif

